JASP-Engine/JASP/R/binomialtest.R

```diff
@@ -15,224 +15,315 @@
-  if (options[["descriptivesPlots"]] && is.null(descriptPlots))
-		descriptPlots <- .binomialDescriptivesPlot(dataset, options, perform, variables)
+  # Error checking
+  .binomCheckErrors(dataset, options)
 
-	results <- list()
-	results[["binomial"]] <- .BinomialTable(binomResults, options)
+  # Compute the results
+  binomResults <- .binomComputeResults(jaspResults, dataset, options)
 
-	if (! is.null(descriptPlots)) {
-		results[["descriptives"]] <- descriptPlots
-		results[["descriptives"]][["title"]] <- "Descriptive Plots"
-	}
+  # Output tables and plots
+  .binomTableMain(jaspResults, dataset, options, binomResults)
+  .binomContainerPlots(jaspResults, dataset, options, binomResults)
+  .binomPlotsDescriptive(jaspResults, dataset, options, binomResults)
 
-  if (perform == "init")
-		return(list(results=results, status="inited", state=state))
+  return()
 
-	state <- list(
-		options = options,
-		binomResults = binomResults,
-		descriptPlots = descriptPlots
-	)
+}
 
-  return(list(results=results, status="complete", state=state))
+# Preprocessing functions ----
+.binomReadData <- function(dataset, options) {
+  if (!is.null(dataset)) {
+    return(dataset)
+  } else {
+    return(.readDataSetToEnd(columns.as.factor = options$variables))
+  }
 }
 
-.binomialTest <- function(dataset, options, perform, variables) {
+.binomCheckErrors <- function(dataset, options) {
+
+  # Check if results can be computed
+  if (length(options$variables) == 0) .quitAnalysis("Input at least one variable")
+
+  # Error Check 1: Number of levels of the variables
+  .hasErrors(
+    dataset              = dataset,
+    perform              = "run",
+    type                 = "factorLevels",
+    factorLevels.target  = options$variables,
+    factorLevels.amount  = "< 1",
+    exitAnalysisIfErrors = TRUE
+  )
+
+  # Error check 2: 0 observations for a level of a variable
+  for (variable in options$variables) {
+
+    column <- dataset[[.v(variable)]]
+    data   <- column[!is.na(column)]
+    levels <- levels(as.factor(data))
+
+    for (level in levels) {
+      .hasErrors(
+        dataset              = data[data == level],
+        perform              = "run",
+        type                 = "observations",
+        observations.amount  = "< 1",
+        exitAnalysisIfErrors = TRUE
+      )
+    }
+  }
 
-  binomResults <- jasp.data.frame(colnames=c("case", "level", "counts", "total", "proportion", "p", "VovkSellkeMPR", "lowerCI", "upperCI"))
+}
 
-  if (perform == "init" || is.null(variables)) {
+# Results functions ----
+.binomComputeResults <- function(jaspResults, dataset, options) {
+
+  # Take results from state if possible
+  if (!is.null(jaspResults[["stateBinomResults"]])) return(jaspResults[["stateBinomResults"]]$object)
+
+  # This will be the object that we fill with results
+  results <- list()
+
+  # First, we perform precalculation of variables we use throughout the analysis
+  results$spec <- .binomCalcSpecs(dataset, options)
```

Comment:
doesn't this mean that if a variable is called `spec` this will cause an error?

Reply:
Nope, results here is simply a list. Do you forsee any issues with this naming? of course, we can just as well change this.

Reply:
Use compute columns to create a variable called `spec`. Run the analysis, observe that it crashes: 

```
Error in switch(binomResults$spec$hypothesisRec, two.sided = .messages('footnote', : EXPR must be a length 1 vector
```

The problem is that a bit lower you call ` results[[variable]] <- list()` which in this scenario overwrites the information in `results[["spec"]]`.

Reply:
Thanks for noticing this! I've fixed it in the most recent push. I added an extra level to the binomResults list so now it looks somewhat like so:

```
binomResults
├─ spec
|  ├─ hypothesisRec
|  └─ levels
|      └─ ...
└─ binom
   ├─ variable1
   |   └─ ...
   └─ variable2
       └─ ...

```